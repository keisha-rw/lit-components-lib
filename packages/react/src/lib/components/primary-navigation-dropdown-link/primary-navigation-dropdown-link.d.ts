import * as React from 'react';
export declare const PdsPrimaryNavigationDropdownLink: React.FunctionComponent<Partial<{
    arrow: boolean;
    variant: "inverted" | "default";
    href: string | undefined;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{
    onClick: (e: Event) => void;
}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
