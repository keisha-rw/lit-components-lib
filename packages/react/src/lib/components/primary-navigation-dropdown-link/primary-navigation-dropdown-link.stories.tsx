import * as React from 'react';
import { action } from '@storybook/addon-actions';
import type { StoryFn } from '@storybook/react';
import { PdsPrimaryNavigationDropdownLink } from './primary-navigation-dropdown-link';

export default {
  title:
    'Components/Primary navigation/Primary navigation main menu/Primary navigation main menu item/Primary navigation dropdown link',
  component: PdsPrimaryNavigationDropdownLink,
  tags: ['autodocs'],
  parameters: {
    happo: false,
    docs: {
      description: {
        component:
          'Note: This component is ONLY meant to be used as a subcomponent of the primary navigation component. It should always be used within a primary-navigation-main-menu-item element.',
      },
    },
  },
};

const Template: StoryFn<typeof PdsPrimaryNavigationDropdownLink> = (args) => (
  <PdsPrimaryNavigationDropdownLink
    {...args}
    onClick={(e: any) =>
      action('pds-primary-navigation-dropdown-link-click')(e.detail)
    }
    href="#"
  >
    This is a link
  </PdsPrimaryNavigationDropdownLink>
);

export const Default = Template.bind({});
Default.args = { variant: 'default' };

export const Inverted = Template.bind({});
Inverted.args = { variant: 'inverted' };
Inverted.parameters = {
  backgrounds: {
    default: 'BrandXStrong',
  },
  viewport: {
    defaultViewport: 'iphonex',
  },
};
