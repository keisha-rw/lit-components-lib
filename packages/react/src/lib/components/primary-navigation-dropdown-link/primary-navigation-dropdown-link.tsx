import { EventName } from '@lit-labs/react';
import { PdsPrimaryNavigationDropdownLink as PdsPrimaryNavigationDropdownLinkElement } from '@keisha/design-system';
import * as React from 'react';
import { PdsPrimaryNavigationDropdownLinkProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsPrimaryNavigationDropdownLink =
  createPdsReactComponent<PdsPrimaryNavigationDropdownLinkProps>()({
    tagName: 'pds-primary-navigation-dropdown-link',
    elementClass: PdsPrimaryNavigationDropdownLinkElement,
    react: React,
    events: {
      onClick:
        'pds-primary-navigation-dropdown-link-click' as EventName<CustomEvent>,
    },
  });
