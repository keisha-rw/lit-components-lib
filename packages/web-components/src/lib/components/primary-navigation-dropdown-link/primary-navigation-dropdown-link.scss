/*------------------------------------*\
 #NAV_DROPDOWN_LINK
\*------------------------------------*/

pds-icon-arrow-right {
  display: inline-block;
  position: relative;
  top: size(0.25);
  left: size(1);
}

.pds-c-primary-navigation-dropdown-link {
  @include pds-semantic-body-default;

  text-decoration: none;
  cursor: pointer;
  color: var(--pds-component-primary-nav-default-dropdown-link);
  text-underline-offset: size(0.75);

  &:hover {
    pds-icon-arrow-right {
      transform: translateX(size(0.5));
      color: var(--pds-component-primary-nav-default-dropdown-hover-link);
      transition: ease-in 0.3s;
    }
  }
}

.pds-primary-navigation-dropdown-link__content {
  .pds-c-primary-navigation-dropdown-link:hover & {
    color: var(--pds-component-primary-nav-default-dropdown-hover-link);
    text-decoration: underline;
  }
}

@media all and (max-width: $pds-bp-lg) {
  .pds-c-primary-navigation-dropdown-link--inverted {
    color: var(--pds-component-primary-nav-inverted-mobile-link);

    &:hover {
      pds-icon-arrow-right {
        transform: translateX(size(0.5));
        color: var(--pds-component-primary-nav-inverted-mobile-hover-link);
        transition: ease-in 0.3s;
      }
    }
  }

  .pds-primary-navigation-dropdown-link__content {
    .pds-c-primary-navigation-dropdown-link--inverted:hover & {
      color: var(--pds-component-primary-nav-inverted-mobile-hover-link);
      text-decoration: underline;
    }
  }
}

@media all and (min-width: $pds-bp-lg) {
  .pds-c-primary-navigation-dropdown-link--inverted {
    color: var(--pds-component-primary-nav-inverted-dropdown-link);

    &:hover {
      pds-icon-arrow-right {
        transform: translateX(size(0.5));
        color: var(--pds-component-primary-nav-inverted-dropdown-hover-link);
        transition: ease-in 0.3s;
      }
    }
  }

  .pds-primary-navigation-dropdown-link__content {
    .pds-c-primary-navigation-dropdown-link--inverted:hover & {
      color: var(--pds-component-primary-nav-inverted-dropdown-hover-link);
      text-decoration: underline;
    }
  }
}
