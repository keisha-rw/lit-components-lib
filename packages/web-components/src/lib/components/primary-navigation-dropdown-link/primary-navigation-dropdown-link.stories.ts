import { html } from 'lit';
import { Meta, StoryObj } from '@storybook/web-components';
import './primary-navigation-dropdown-link';

const meta: Meta = {
  title:
    'Components/Primary navigation/Primary navigation main menu/Primary navigation main menu item/Primary navigation dropdown link',
  component: 'pds-primary-navigation-dropdown-link',
  tags: ['autodocs'],
  parameters: {
    happo: false,
    actions: {
      handles: ['pds-primary-navigation-dropdown-link-click'],
    },
    docs: {
      description: {
        component:
          'Note: This component is ONLY meant to be used as a subcomponent of the primary navigation component. It should always be used within a primary-navigation-main-menu-item element.',
      },
    },
  },
};
export default meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Default: StoryObj = {
  name: 'Default',
  render: () =>
    html`<pds-primary-navigation-dropdown-link href="#"
      >This is a link</pds-primary-navigation-dropdown-link
    >`,
};

export const Inverted: StoryObj = {
  name: 'Inverted',
  parameters: {
    backgrounds: {
      default: 'BrandXStrong',
    },
    viewport: {
      defaultViewport: 'iphonex',
    },
    a11y: {
      config: {
        rules: [
          {
            // Color contrast is sufficient for the inverted links.
            //
            // @link https://dequeuniversity.com/rules/axe/4.6/color-contrast
            id: 'color-contrast',
            enabled: false,
          },
        ],
      },
    },
  },
  render: () =>
    html`<pds-primary-navigation-dropdown-link href="#" variant="inverted"
      >This is a link</pds-primary-navigation-dropdown-link
    >`,
};
