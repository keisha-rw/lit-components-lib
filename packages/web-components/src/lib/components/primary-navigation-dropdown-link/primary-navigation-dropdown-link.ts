import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './primary-navigation-dropdown-link.scss?inline';
import '@keisha/design-system-icons-web/arrow-right';

/**
 * @summary This component renders a styled anchor tag for the primary-nav
 *
 * @slot default This slot holds the link contents
 *
 * @fires pds-primary-navigation-dropdown-link-click A custom event dispatched on click
 */
@customElement('pds-primary-navigation-dropdown-link', {
  category: 'component',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsPrimaryNavigationDropdownLink extends PdsElement {
  /**
   * The link for the item
   */
  @property({ type: String })
  href?: string;

  /**
   * Determines if the item should have an arrow icon
   */
  @property({ type: Boolean })
  arrow: boolean;

  /**
   * Style variant
   * - **default** renders the default primary-navigation
   * - **inverted** renders the inverted primary-navigation
   */
  @property()
  variant: 'default' | 'inverted' = 'default';

  handleClick() {
    const customEvent = new CustomEvent(
      'pds-primary-navigation-dropdown-link-click',
      {
        bubbles: true,
        composed: true,
        detail: {
          summary: this.textContent,
        },
      },
    );

    this.dispatchEvent(customEvent);
  }

  /**
   * @internal
   */
  get classNames() {
    return {
      arrow: !!this.arrow,
      [this.variant]: !!this.variant,
    };
  }

  render() {
    return html`<a
      href=${this.href}
      @click=${this.handleClick}
      class=${this.getClass()}
    >
      <span class="pds-primary-navigation-dropdown-link__content"
        ><slot>This is a link</slot></span
      >
      ${this.arrow === true
        ? html`<pds-icon-arrow-right size="sm"></pds-icon-arrow-right>`
        : ''}
    </a>`;
  }
}
