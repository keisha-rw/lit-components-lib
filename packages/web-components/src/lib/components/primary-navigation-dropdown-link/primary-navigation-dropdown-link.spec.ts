import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsPrimaryNavigationDropdownLink } from './primary-navigation-dropdown-link';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Primary navigation dropdown link/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('PrimaryNavigationDropdownLink unit tests', () => {
  let element: Element;

  beforeEach(async () => {
    element = await fixture(
      '<pds-primary-navigation-dropdown-link target="_blank" href="https://www.google.com"><span>This is a link</span></pds-primary-navigation-dropdown-link>',
    );
  });

  it('is an instance of PdsPrimaryNavigationDropdownLink', () => {
    expect(element).toBeInstanceOf(PdsPrimaryNavigationDropdownLink);
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('dispatches a custom event on click and gets correct summary', async () => {
    const linkEl = element.shadowRoot?.querySelector('a');

    const handler = jest.fn();
    element.addEventListener(
      'pds-primary-navigation-dropdown-link-click',
      handler,
    );

    linkEl?.dispatchEvent(new Event('click'));

    return Promise.resolve().then(() => {
      expect(handler.mock.calls[0][0].detail.summary).toEqual('This is a link');
    });
  });

  it('populates the arrow icon when provided', async () => {
    element = await fixture(
      '<pds-primary-navigation-dropdown-link arrow>This is a link</pds-primary-navigation-dropdown-link>',
    );
    expect(element).toMatchSnapshot('This is a link');
  });
});
