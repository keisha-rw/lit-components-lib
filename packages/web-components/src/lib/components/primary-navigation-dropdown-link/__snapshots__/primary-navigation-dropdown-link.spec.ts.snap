// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryNavigationDropdownLink unit tests populates the arrow icon when provided: This is a link 1`] = `
<pds-primary-navigation-dropdown-link
  arrow=""
>
  This is a link
</pds-primary-navigation-dropdown-link>
`;
